include/dg/PointerAnalysis/PSNode.h

```diff
@@ -744,6 +706,76 @@ class PSNodeJoin : public PSNode {
     friend class PSNodeFork;
 };
 
+template<PSNodeType Type, typename = void>
+struct GetNodeType
+{
+    using type = PSNode;
+};
+
+template<>
+struct GetNodeType<PSNodeType::ALLOC>
+{
+    using type = PSNodeAlloc;
+};
+
+template<>
+struct GetNodeType<PSNodeType::CONSTANT>
+{
+    using type = PSNodeConstant;
+};
+
+template<>
+struct GetNodeType<PSNodeType::GEP>
+{
+    using type = PSNodeGep;
+};
+
+template<>
+struct GetNodeType<PSNodeType::MEMCPY>
+{
+    using type = PSNodeMemcpy;
+};
+
+
+template<>
+struct GetNodeType<PSNodeType::ENTRY>
+{
+    using type = PSNodeEntry;
+};
+
+template<PSNodeType Type>
+struct GetNodeType<Type,
+    typename std::enable_if<Type == PSNodeType::CALL || Type == PSNodeType::CALL_FUNCPTR>
+                           ::type>
+{
+    using type = PSNodeCall;
+};
+
+template<>
+struct GetNodeType<PSNodeType::FORK>
+{
+    using type = PSNodeFork;
+};
+
+template<>
+struct GetNodeType<PSNodeType::JOIN>
+{
+    using type = PSNodeJoin;
+};
+
+template<>
+struct GetNodeType<PSNodeType::RETURN>
+{
+    using type = PSNodeRet;
+};
+
+template<>
+struct GetNodeType<PSNodeType::CALL_RETURN>
+{
+    using type = PSNodeCallRet;
+};
+
+
```

Comment:
C++17 note: This whole part can be replaced with `if constexpr` in `dg::pta::PointerGraph::nodeFactory`